File: 백준/Gold/1806. 부분합/부분합.py
# 수열에서 연속된 수들의 부분합 중에 합이 S 이상이 되는 것 중 가장 짧은 것의 길이를 구하는 문제
# 누적합을 구한 후 투포인터 사용
N, S = map(int, input().split())        # 길이, 합
lst = list(map(int, input().split()))
cumulative_sum = [0] * (N+1)
answer = N
start, end = 0, 1

if sum(lst) < S :       # 불가능
    print(0)
else :
    # 누적합 구하기
    for n in range(1, N+1):
        cumulative_sum[n] = cumulative_sum[n - 1] + lst[n-1]

    # 투포인터
    while start != N :
        if cumulative_sum[end] - cumulative_sum[start] >= S :
            answer = min(answer, end - start)
            start += 1
        else :      # 합이 S보다 작으면 오른쪽으로 이동
            if end != N :
                end += 1
            else :      # end가 끝에 도달해서 start를 움직인다
                start += 1

    print(answer)

File: 백준/Gold/1806. 부분합/부분합.java
import java.io.*;

class Main {
    static int s;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] firstInput = br.readLine().split(" ");
        int n = Integer.parseInt(firstInput[0]);        // 10 <= 수열의 길이 < 100,000
        s = Integer.parseInt(firstInput[1]);        // 0 < 합 <= 100,000,000
        String[] secondInput = br.readLine().split(" ");
        int[] arr = new int[n + 1];        // 누적합 배열
        int answer = Integer.MAX_VALUE;   // 최소 길이 초기화
        for(int i = 0; i < n; i++) {
            int num = Integer.parseInt(secondInput[i]);
            arr[i + 1] = arr[i] + num;    // i+1번째 값까지의 누적합
        }

        // 투포인터 알고리즘 실행
        if(arr[n] >= s) {  // 마지막 누적합 값이 합 S 이상이면
            answer = twoPointerAlgo(arr);
        }

        // answer가 갱신되지 않았다면 만들 수 있는 부분합이 없음
        if (answer == Integer.MAX_VALUE) {
            answer = 0;
        }

        System.out.print(answer);
    }

    /*
        arr: 누적합 배열
    */
    public static int twoPointerAlgo(int[] arr) {
        int left = 0;
        int right = 1;
        int result = arr.length;        // 최소 길이
        while(right < arr.length) {    
            if(arr[right] - arr[left] >= s) {    // 합이 S 이상이면
                result = Math.min(result, right - left);    // 최소 길이 갱신
                left++;    // 범위를 줄이기 위해 왼쪽 포인터의 인덱스를 증가
            } else {
                right++;    // 합이 S보다 작아서 오른쪽 포인터의 인덱스를 증가
            }
        }
        return result;
    }
}
